Mathlib/Algebra/DirectSum/Internal.lean

```diff
@@ -276,6 +276,57 @@ theorem coe_of_mul_apply {i : ι} (r : A i) (r' : ⨁ i, A i) (n : ι) [Decidabl
 
 end CanonicallyOrderedAddCommMonoid
 
+section GradeZeroSubsemiring
+
+variable [Semiring R]
+variable [CanonicallyOrderedAddCommMonoid ι]
+variable [SetLike σ R] [AddSubmonoidClass σ R]
+variable (A : ι → σ) [SetLike.GradedMonoid (fun i ↦ A i)]
+
+/--
+The subsemiring `A 0` of `R`.
+-/
+def GradeZero.subsemiring : Subsemiring R where
+  carrier := A 0
+  mul_mem' := by
+    intro a b ha hb
```

Comment:
```suggestion
  mul_mem' a b ha hb := by
```

Reply:
No this doesn't work. I have tried to change my codes according to what you said, but it resulted in an error:
```
type mismatch
  fun a b ha hb ↦ ?m.460700 a b ha hb
has type
  (a : a✝ ∈ ↑(A 0)) →
    (b : b✝ ∈ ↑(A 0)) →
      (ha : ?m.460697 a b) → (hb : ?m.460698 a b ha) → ?m.460699 a b ha hb : Sort (imax ?u.460688 ?u.460691 ?u.460694)
but is expected to have type
  a✝ ∈ ↑(A 0) → b✝ ∈ ↑(A 0) → a✝ * b✝ ∈ ↑(A 0) : Prop
the following variables have been introduced by the implicit lambda feature
  a✝ : R
  b✝ : R
you can disable implicit lambdas using `@` or writing a lambda expression with `{}` or `[]` binder annotations.
```

Reply:
looks like a and b are implicit -- try `mul_mem' ha hb := by`?

Reply:
Or alternatively, `mul_mem' {a b} ha hb := by`, as the error message is alluding to.

Reply:
Yeah now it works! Thank you very much!